src/editor/nodes/ParticleNode.js

```diff
@@ -0,0 +1,416 @@
+        velFactor[i] = 0;
+        continue;
+      }
+
+      if (this.ages[i] > this.lifetimes[i]) {
+        position[i * 4] = this.initialPositions[i * 3];
+        position[i * 4 + 1] = this.initialPositions[i * 3 + 1];
+        position[i * 4 + 2] = -2; //this.initialPositions[i * 3 + 2];
+        velFactor[i] = 0;
+        this.ages[i] = this.initialAges[i];
+        colorFactor[i] = 0;
+
+        continue;
+      }
+
+      switch (this.velocityCurve) {
+        case "Linear":
+          break;
+        case "Ease-in":
+          velFactor[i] = this.EaseIn(velFactor[i]);
+          break;
+        case "Ease-out":
+          velFactor[i] = this.EaseOut(velFactor[i]);
+          break;
+        case "Ease-in, out":
+          velFactor[i] = this.EaseInOut(velFactor[i]);
+          break;
+      }
+
+      this.velocities[i * 3] = this.lerp(this.velocity.x, this.endVelocity.x, velFactor[i]);
+      this.velocities[i * 3 + 1] = this.lerp(this.velocity.y, this.endVelocity.y, velFactor[i]);
+      this.velocities[i * 3 + 2] = this.lerp(this.velocity.z, this.endVelocity.z, velFactor[i]);
+
+      position[i * 4] += this.velocities[i * 3] * dt;
+      position[i * 4 + 1] += this.velocities[i * 3 + 1] * dt;
+      position[i * 4 + 2] += this.velocities[i * 3 + 2] * dt;
+
+      angleVel = this.angularVelocity;
+      customAngle[i] += angleVel * 0.01745329251 * dt;
+
+      switch (this.colorCurve) {
+        case "Even":
+          colorFactor[i] = this.Even(colorFactor[i]);
+          break;
+        case "Ease-in":
+          colorFactor[i] = this.EaseIn(colorFactor[i]);
+          break;
+        case "Ease-out":
+          colorFactor[i] = this.EaseOut(colorFactor[i]);
+          break;
+        case "Ease-in, out":
+          colorFactor[i] = this.EaseInOut(colorFactor[i]);
+          break;
+      }
+
+      if (colorFactor[i] <= 0.5) {
+        color[i * 4] = this.lerp(this.startColor.r, this.middleColor.r, colorFactor[i] / 0.5);
+        color[i * 4 + 1] = this.lerp(this.startColor.g, this.middleColor.g, colorFactor[i] / 0.5);
+        color[i * 4 + 2] = this.lerp(this.startColor.b, this.middleColor.b, colorFactor[i] / 0.5);
+        color[i * 4 + 3] = this.lerp(this.startOpacity, this.middleOpacity, colorFactor[i] / 0.5);
+      } else if (colorFactor[i] > 0.5) {
+        color[i * 4] = this.lerp(this.middleColor.r, this.endColor.r, (colorFactor[i] - 0.5) / 0.5);
+        color[i * 4 + 1] = this.lerp(this.middleColor.g, this.endColor.g, (colorFactor[i] - 0.5) / 0.5);
+        color[i * 4 + 2] = this.lerp(this.middleColor.b, this.endColor.b, (colorFactor[i] - 0.5) / 0.5);
+        color[i * 4 + 3] = this.lerp(this.middleOpacity, this.endOpacity, (colorFactor[i] - 0.5) / 0.5);
+      }
+
+      this.geometry.attributes.position.needsUpdate = true;
+      this.geometry.attributes.color.needsUpdate = true;
+      this.geometry.attributes.customAngle.needsUpdate = true;
+    }
+  }
+
+  lerp(start, end, a) {
```

Comment:
I'd move all of these helper functions to the top of the file to avoid using `this.lerp` etc.